scripts/swo_parser.py

```diff
@@ -0,0 +1,182 @@
+            self.tcl_socket.sendall(b'puts "' + s.encode('utf-8') + b'"\r\n\x1a')
+        
+        
+class StreamManager:
+    """
+    Manages up to 32 byte streams.
+    
+    This class contains a dictionary of streams indexed by their stream id. It
+    is responsible for parsing the incoming data, and forwarding the bytes to
+    the correct stream.
+    
+    """
+    def __init__(self):
+        self.streams = dict()
+        self._itmbuffer = b''
+        
+    def add_stream(self, stream):
+        self.streams[stream.id] = stream
+        
+    def parse_tcl(self, line):
+        r"""
+        When OpenOCD is configured to output the trace data over the Tcl
+        server, it periodically outputs a string (terminated with \r\n) that
+        looks something like this:
+        
+        type target_trace data 01480165016c016c016f0120015401720161016301650121010a
+        
+        The parse_tcl method turns this into the raw ITM bytes and sends it to
+        parse_itm_bytes.
+        
+        """
+        if (line.startswith(b'type target_trace data ') and 
+            line.endswith(b'\r\n')
+            ):
+            itm_bytes = int(line[23:-2],16).to_bytes(len(line[23:-2])//2,
+                                               byteorder='big')
+            self.parse_itm_bytes(itm_bytes)
+                    
+    def parse_itm_bytes(self, bstring):
+        """
+        Parses ITM packets based on the format discription from ARM
+        http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0314h/Chdbicbg.html
+        
+        """
+        
+        bstring = self._itmbuffer + bstring
+        self._itmbuffer = b''
+        
+        while len(bstring) > 0:
+            header = bstring[0]
+            # The third bit of a header must be zero, and the last two bits
+            # can't be zero.
+            if header & 0x04 != 0 or header & 0x03 == 0:
+                bstring = bstring[1:]
+                continue
+                                
+            payload_size = 2**(header & 0x03 - 1)
+            stream_id = header >> 3
+            
+            if payload_size >= len(bstring):
+                self._itmbuffer = bstring
+                return
+                
+            if stream_id in self.streams:
+                s = bstring[1:payload_size+1].decode('ascii')
+                self.streams[stream_id].add_chars(s)
+            
+            bstring = bstring[payload_size+1:]
+   
+
+#### Main program ####
+
+# Set up the socket to the OpenOCD Tcl server
+HOST = 'localhost'
+PORT = 6666
+with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as tcl_socket:
+    tcl_socket.connect((HOST, PORT))
+    tcl_socket.settimeout(0)
+    
+    # Create a stream manager and add three streams
+    streams = StreamManager()
+    streams.add_stream(Stream(0, '', tcl_socket))
+    streams.add_stream(Stream(1, 'WARNING: '))
+    streams.add_stream(Stream(2, 'ERROR: ', tcl_socket))
```

Comment:
this is a cool idea we should use when we get a chance to improve our logger. currently we only send over port0

Reply:
interesting, it's able to log to different ports?

Reply:
yep